### DISCUSSION TITLE
Is it possible to mimic QAT (quantization-aware training) "by hand" by customizing a model's forward pass?

### DISCUSSION THREAD

**Discussion by incf-ai** [tracel-ai/burn#5107] [Q&A] [answered]
Hi,

I've been playing around with Burn to educate myself as I work through my master degree (I have huge gaps in my knowledge so bear with me). I'm interested in using Burn to develop models that run on slow embedded NPUs that only support integer math (INT8 only is fairly common on small "low power consumption" NPUs).

From what I've seen with other types of models QAT (quantization-aware training) is significantly better than PTQ (post-training quantization) and on the embedded device's I'm targeting I expect that it would make a large difference in terms of helping me reduce parameter count and etc.

The Burn documentation gives the briefest hint of how it's normally done, but also states it's unsupported - which I take to mean it's unimplemented currently (that's fine).  If I was sufficiently motivated and knowledgeable, is it possible to mimic QAT just by customizing a model's forward pass during training? (for example if I were to modify the model definition in the MNIST CNN training example)

There is a lot that I don't know, I apologize if this is a silly question.

edit: after doing just a little reading I think I see it's not so simple with portions of `burn-autodiff/src/ops/qtensor.rs` requiring implementation and testing

**Comment by Mikyx-1** [comment, marked as answer]
@incf-ai 
Your edit is exactly right, and yes — you can absolutely mimic QAT by hand, without touching `QTensorOps`/`burn-autodiff/src/ops/qtensor.rs` at all.

The key insight is that "QAT" in most frameworks (including PyTorch) isn't actually running int8 matmuls during training — it's *fake quantization*: you round-trip float tensors through a simulated quantize→dequantize step in the forward pass so the model "feels" the discretization error, while gradients still flow in float. That's a pure `Tensor<B, D>` (float) technique, so it doesn't depend on Burn's native `QuantizedTensor` machinery being finished (the `todo!()` you found in `qtensor.rs` is only needed if you want *real* packed-int8 storage + autodiff through it, which is a different, harder problem than what QAT itself requires).

The one wrinkle is that `round()`/`clamp()` have zero gradient almost everywhere, which would kill backprop. The standard fix is the **straight-through estimator (STE)**: use the quantized value in the forward pass, but let the gradient pass through as if quantization were the identity function. In Burn you can build this with `.detach()` (in `burn_tensor::Tensor`, `crates/burn-tensor/src/tensor/api/float.rs`):

```rust
fn fake_quantize<B: Backend, const D: usize>(
    x: Tensor<B, D>,
    scale: f32,
    qmin: f32,
    qmax: f32,
) -> Tensor<B, D> {
    let x_detached = x.clone().detach();

    // simulate int8 round-trip while staying in f32
    let x_q = (x_detached.clone() / scale)
        .round()
        .clamp(qmin, qmax)
        * scale;

    // straight-through estimator:
    // forward value == x_q, but d(out)/dx == 1 (identity), since the
    // (x_q - x_detached) term is constant w.r.t. the graph and only
    // `x` itself carries gradient tracking.
    x + (x_q - x_detached)
}
```

Call `fake_quantize` on weights and/or activations inside your model's `forward()` (e.g. right after each `Linear`/`Conv2d` in the MNIST CNN example), and everything else — optimizer, backward, loss — works unmodified because the output is still a plain float tensor.

A few practical notes:
- You can reuse Burn's existing calibration/scale-computation helpers (`compute_range`, `compute_q_params` in `burn_tensor::quantization`, the same ones `Quantizer`/`quantize_weights` use for PTQ in `crates/burn-core/src/module/quantize.rs`) instead of hand-rolling scale computation — that keeps you consistent with however you eventually do PTQ export.
- Per-channel or per-block scales just mean computing `scale` per output channel and broadcasting instead of a single scalar.
- This gets you the training-time benefit of QAT (the model learns to be robust to quantization noise), but note it does **not** give you an actual quantized, int8-executing model at the end — for deployment you'd still export via Burn's PTQ path (`model.quantize_weights(...)`) using the ranges your fake-quantized model already learned to be robust to, or hand off the calibrated scales/weights to your NPU toolchain's own int8 converter.
- This is essentially the same trick used in `torch.ao.quantization`'s `FakeQuantize` module and the original QAT paper (Jacob et al., 2017) — nothing Burn-specific is missing to replicate it.

So: not a silly question at all, and no, you don't need to wait for `QTensorOps::quantize` to land in `burn-autodiff` to experiment with QAT today.

**Reply by nathanielsimard** [reply]
That's a great response! We're working on automatically supporting LoRA and QLoRA in Burn. That means we're working on improving the Param struct so that you don't have to change the model code to be able to use LoRA. We could use the same technique for QAT.
